External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp

```diff
@@ -0,0 +1,154 @@
+#include "Interface/IR/PassManager.h"
+#include "Interface/Core/OpcodeDispatcher.h"
+
+namespace FEXCore::IR {
+
+class DeadGPRStoreElimination final : public FEXCore::IR::Pass {
+public:
+  bool Run(IREmitter *IREmit) override;
+};
+
+struct GPRInfo {
+  uint32_t reads { 0 };
+  uint32_t writes { 0 };
+  uint32_t kill { 0 };
+};
+
+bool IsFullGPR(uint32_t Offset, uint8_t Size) {
+  if (Size != 8)
+    return false;
+  if (Offset & 7)
+    return false;
+  
+  if (Offset < 8 || Offset >= (17 * 8))
+    return false;
+
+  return true;
+}
+
+bool IsGPR(uint32_t Offset) {
+  
+  if (Offset < 8 || Offset >= (17 * 8))
+    return false;
+
+  return true;
+}
+
+uint32_t GPRBit(uint32_t Offset) {
+  if (!IsGPR(Offset)) {
+    return 0;
+  }
+
+  return  1 << ((Offset - 8)/8); 
+}
+
+/**
+ * @brief This is a temporary pass to detect simple multiblock dead GPR stores
+ *
+ * First pass computes which GPRs are read and written per block
+ *
+ * Second pass computes which GPRs are stored, but overwritten by the next block(s).
+ * It also propagates this information a few times to catch dead GPRs across multiple blocks.
+ *
+ * Third pass removes the dead stores.
+ *
+ */
+bool DeadGPRStoreElimination::Run(IREmitter *IREmit) {
+  std::map<OrderedNode*, GPRInfo> GPRMap;
+
+  bool Changed = false;
+  auto CurrentIR = IREmit->ViewIR();
+
+  // Pass 1
+  // Compute GPRs read/writes per block
+  // This is conservative and doesn't try to be smart about loads after writes
+  {
+    for (auto [BlockNode, BlockIROp] : CurrentIR.GetBlocks()) {
+      for (auto [CodeNode, IROp] : CurrentIR.GetCode(BlockNode)) {
+
+        if (IROp->Op == OP_STORECONTEXT) {
+          auto Op = IROp->CW<IR::IROp_StoreContext>();
+          if (IsFullGPR(Op->Offset, IROp->Size))
+            GPRMap[BlockNode].writes |= GPRBit(Op->Offset);
+          else
+            GPRMap[BlockNode].reads |= GPRBit(Op->Offset);
+        }
+        else if (IROp->Op == OP_STORECONTEXTINDEXED ||
+               IROp->Op == OP_LOADCONTEXTINDEXED ||
+               IROp->Op == OP_LOADCONTEXTPAIR ||
+               IROp->Op == OP_STORECONTEXTPAIR) {
```

Comment:
We also can't track through syscall here.
Syscall can modify full context state...
We should also probably move rax setting from syscall to be handled by syscall handler so we are more generic...

Reply:
RCLSE, DeadFlags and RA assume that SYSCALLS don't modify the context. Do syscalls really do it?

Reply:
Will resolve later